src/json/def_to_json.cr

```diff
@@ -0,0 +1,177 @@
+module JSON
+  # The `def_to_json` macro generates a `to_json(json : JSON::Builder)` method based on provided *mappings*.
+  #
+  # It is a lightweight alternative to `JSON.mapping` if you don't need to declare instance variables and a parser.
+  #
+  # The generated method invoks `to_json(JSON::Builder)` on each of the values returned
+  # by `self`'s properties from *mappings*, or - if a converter is specified - `to_json(value, JSON::Builder)` on the converter.
+  #
+  # ### Example
+  #
+  # ```
+  # require "json"
+  #
+  # record Location, lat : Float64, long : Float64 do
+  #   JSON.def_to_json([lat, long])
+  # end
+  #
+  # record House, street : String, street_number : Int32, location : Location do
+  #   JSON.def_to_json(
+  #     address: true,
+  #     loc: {property: location},
+  #     empty_field: {emit_null: true},
+  #   )
+  #
+  #   def address
+  #     "#{street} #{street_number}"
+  #   end
+  #
+  #   def empty_field
+  #     nil
+  #   end
+  # end
+  #
+  # house = House.new("Crystal Road", 1234, Location.new(12.3, 34.5))
+  # house.to_json # => %({"address":"Crystal Road 1234","loc":{"lat":12.3,"long":34.5},"empty_field":null})
+  # ```
+  #
+  # ### Usage
+  #
+  # `JSON.def_to_json` must receive a series of named arguments, or a named tuple literal, or a hash literal,
+  # whose keys will define JSON properties.
+  #
+  # The value of each key can either be `true` or a hash or named tuple literal with the following options:
+  # * **property**: the property name on the Crystal object (as opposed to the key in the JSON document)
```

Comment:
I think it makes more sense this way: with `def_to_json` you're basically describing a JSON object schema. The keys of this schema correspond to the keys of the named tuple and the value defines which value is to be assigned.
`mapping` on the other hand has a different focus: it primarily defines properties as instance variables and adds methods to convert them to JSON and back. Here it makes sense to have the crystal properties as keys (because that's the main part) and the JSON field names as arguments.
While the `properties` argument looks quite similar, both macros have a different usecase and allow different options, so I don't think it is bad to have different semantics between key and value.

For `def_to_json` there is also the benefit that this way it is very easy to use arbitrary expressions as value:
```crystal
@name = "FooBar"
JSON.def_to_json({
  name: { property: name.downcase }
})
# to_json will output {"name":"foobar"}
```
This would not be possible or look very nasty if it was used as a key xD